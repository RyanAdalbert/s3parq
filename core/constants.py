import os
import yaml
from .helpers.project_root import ProjectRoot
from git import Repo

## IMPORTANT: config/core_project.yaml is in core's .gitignore, be sure to remove it before making
## any changes to constants

## DYNAMICS
# AWS_ACCOUNT default is set to PROD_AWS_ACCOUNT for prod and uat, DEV_AWS_ACCOUNT otherwise
# BATCH_JOB_QUEUE default is set to '<ENVIRONMENT>-core'
# ENV_BUCKET default is set to 'ichain-<ENVIRONMENT>' 

def reset_constants():
    ''' sets (or resets) constants based on hierarchy of 
        1) current envars that begin with ICHAIN_
        2) values defined in core_project.yml
    '''
    config_file = os.path.join(
        ProjectRoot().get_path(), 'config', 'core_project.yaml')
    with open(config_file) as _file:
        config = yaml.safe_load(_file)
        for k, v in config.items():
            globals()[k] = v

    # set constants from all envars starting with ICHAIN_. Override those set in the yaml file
    for key in os.environ.keys():
        if key[:7] == 'ICHAIN_':
            globals()[key.replace('ICHAIN_', '')] = os.environ[key]

    # set the branch name based on the current repo branch, and account for Jenkins weirdness with detached heads
    def get_branch_name():
        # if we already picked up branch name from ICHAIN_BRANCH_NAME, leave it
        if 'ICHAIN_BRANCH_NAME' in os.environ.keys():
            return os.environ["ICHAIN_BRANCH_NAME"]
        # if dev, we need to use the branch name or the os environment variable for jenkins
        elif globals()['ENVIRONMENT'] == 'dev':
            try:
                repo = Repo(ProjectRoot().get_path())
                return repo.active_branch.name

            # jenkins weirdness in dev
            except:
                return os.environ['BRANCH_NAME']
        # for prod and uat if they are hard-coded? for whatever reason, use the environment name
        else:
            return globals()['ENVIRONMENT']

    def get_env_bucket():
        if 'ICHAIN_ENV_BUCKET' in os.environ.keys():
            return os.environ['ICHAIN_ENV_BUCKET']
        else:
            return f"ichain-{globals()['ENVIRONMENT']}"

    def get_aws_account():
        if 'ICHAIN_AWS_ACCOUNT' in os.environ.keys():
            return os.environ['ICHAIN_AWS_ACCOUNT']
        elif globals()['ENVIRONMENT'] in ('prod', 'uat'):
            return globals()['PROD_AWS_ACCOUNT']
        else:
            return globals()['DEV_AWS_ACCOUNT']

    def get_batch_job_queue():
        if 'ICHAIN_BATCH_JOB_QUEUE' in os.environ.keys():
            return os.environ['ICHAIN_BATCH_JOB_QUEUE']
        else:
            return f"{globals()['ENVIRONMENT']}-core"

    def get_core_job_def_name():
        if 'ICHAIN_BATCH_JOB_DEFINITION_NAME' in os.environ.keys():
            return os.environ['ICHAIN_BATCH_JOB_DEFINITION_NAME']
        else:
            if ENVIRONMENT == 'dev':
                return f"core_{BRANCH_NAME}"
            elif ENVIRONMENT == 'uat':
                return "core_uat"
            elif ENVIRONMENT == 'prod':
                return "core_prod"    
            else:
                raise Exception(f"Can't create a core tag job definition name for environment {ENVIRONMENT}")

    # Dynamic (smart) Constants
    globals()['BRANCH_NAME'] = get_branch_name()
    globals()['ENV_BUCKET'] = get_env_bucket()
    globals()['AWS_ACCOUNT'] = get_aws_account()
    globals()['BATCH_JOB_QUEUE'] = get_batch_job_queue()
    globals()['BATCH_JOB_DEFINITION_NAME'] = get_core_job_def_name()

reset_constants()
