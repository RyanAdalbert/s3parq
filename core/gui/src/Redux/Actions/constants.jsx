export const GOOGLE_LOGIN_RESPONSE = 'GOOGLE_LOGIN_RESPONSE';
