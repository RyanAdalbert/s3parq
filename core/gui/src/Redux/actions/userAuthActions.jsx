import history from '../../utils/history';

//User Constants
export const userConstants = {
  STORE_TOKEN: 'STORE_TOKEN',
  LOGIN_ATTEMPT: 'LOGIN_ATTEMPT',
  LOGIN_FAIL: 'LOGIN_FAIL',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  USER_LOGOUT: 'USER_LOGOUT'
};

// store token action creator
export const storeToken = (oAuthToken, userName) => ({
  type: 'STORE_TOKEN',
  payload: {
    oAuthToken,
    userName
  }
});

// login flow action creators
export const loginError = error => {
  return {
    type: 'LOGIN_FAIL',
    error
  };
};

export const loginSuccess = response => {
  return dispatch => {
    dispatch({
      type: 'LOGIN_SUCCESS',
      response
    });
    history.push('/admin');
  };
};

export const loginRequest = oAuthToken => {
  return {
    type: 'LOGIN_ATTEMPT',
    oAuthToken
  };
};

//login action creator
export const login = userData => {
  const HOST = 'localhost';
  return dispatch =>
    fetch(`http://${HOST}:5000/config_api/login`, {
      method: 'GET',
      headers: {
        Authorization: userData.oAuthToken
      }
    })
      .then(response => {
        if (response.status === 200) {
          console.log(response);
          dispatch(loginSuccess(response));
        } else {
          const error = new Error(response.statusText);
          error.response = response;
          dispatch(loginError(error));
          throw error;
        }
      })
      .catch(error => {
        console.log('request failed', error);
      });
};

export const logOut = () => {
  return {
    type: 'USER_LOGOUT'
  };
};
