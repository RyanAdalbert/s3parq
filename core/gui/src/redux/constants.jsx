export const API_HOST = 'https://localhost:5000';
