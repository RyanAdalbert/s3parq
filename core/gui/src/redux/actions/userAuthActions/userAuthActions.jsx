import history from '../../../utils/history';
import { API_HOST } from '../../constants.jsx';

//User Constants
export const userConstants = {
  STORE_TOKEN: 'STORE_TOKEN',
  LOGIN_ATTEMPT: 'LOGIN_ATTEMPT',
  LOGIN_FAIL: 'LOGIN_FAIL',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  USER_LOGOUT: 'USER_LOGOUT'
};

export const loginAttempt = oAuthToken => {
  return {
    type: 'LOGIN_ATTEMPT',
    oAuthToken
  };
};

// login flow action creators
export const loginError = error => {
  return {
    type: 'LOGIN_FAIL',
    error
  };
};

export const loginSuccess = response => {
  return dispatch => {
    dispatch({
      type: 'LOGIN_SUCCESS',
      response
    });
    history.push('/admin/pipelineindex');
  };
};

// store token action creator
export const storeToken = (oAuthToken, userName) => ({
  type: 'STORE_TOKEN',
  payload: {
    oAuthToken,
    userName
  }
});

//login action creator
export const login = oAuthToken => {
  return dispatch => {
    return fetch(`${API_HOST}/config_api/login`, {
      method: 'GET',
      headers: {
        Authorization: oAuthToken
      },
      credentials: 'include'
    })
      .then(response => {
        if (response.status === 200) {
          dispatch(loginSuccess(response));
        } else {
          const error = new Error(response.statusText);
          error.response = response;
          dispatch(loginError(error));
          throw error;
        }
      })
      .catch(error => {
        console.log('request failed', error);
      });
  };
};

//logout action creator
export const logOut = () => ({
  type: 'USER_LOGOUT'
});
