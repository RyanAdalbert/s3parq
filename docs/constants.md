# CONSTANTS

    from core.constants import ENVIRONMENT, AWS_ACCOUNT

## Why? 
A key part of drying out code that is deployed to multiple environments is external context management. CORE takes the kiss approach with `core.constants`

## What Are They? 
each constant exposed by `core.constants` is set in order of:
- `ICHAIN_xxx` environment variable; if you set `ICHAIN_ENV_BUCKET=hamburger` in your environment, it will populate throughout core. 
- Dynamically set constants: we set `ENV_BUCKET`, `AWS_ACCOUNT`, `BRANCH_NAME` and `BATCH_JOB_QUEUE` dynamically based on the value of `ENVIRONMENT`. 
- `config/core_project.yml` defaults - these are fallbacks

## The _BRANCH\_NAME_ Constant
This one is special. It is set:
- by `ICHAIN_BRANCH_NAME` always, first and foremost
- if `ENVIRONMENT` is "dev", core will first attempt to sniff out and set based on the checked out git branch (if there is one). If no git branch is found, it falls back to the `BRANCH_NAME` envar (this is a Jenkins thing).
- if `ENVIRONMENT` is "uat" or "prod", it will be set to that ("uat" or "prod", respectively). 

## The _FORCE\_POSTGRES_ Constant
When `ENVIRONMENT` is dev, the configuration db can be mocked using a SQLite local instance rather than using the local postgres db. If `FORCE_POSTGRES` is true, all calls to the configuration SessionHelper (`/core/helpers/session_helper.py`) will force postgres even in dev. Rather than modifying the constant's value directly in `core_project.yml`, you can import `core.helpers.postgres_toggle` and call either `postgres()` or `cmock()` in any core module to specify the db instance to use in dev.