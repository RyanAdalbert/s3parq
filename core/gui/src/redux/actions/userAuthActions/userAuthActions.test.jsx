import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import expect from 'expect'; // You can use any testing library
import fetchMock from 'fetch-mock';
import { userConstants } from './userAuthActions';
import * as actions from './userAuthActions';

import { API_HOST } from '../../constants';
// const middlewares = [thunk];
// const mockStore = configureMockStore(middlewares);

// describe('async actions', () => {
//   afterEach(() => {
//     // clear all HTTP mocks after each test
//     nock.cleanAll();
//   });

//   it('creates LOGIN_SUCCESS when login authorization gets a 200 response', () => {
//     // Simulate a successful response
//     nock('http://localhost:5000')
//       .get('/config_api/login') // Route to catch and mock
//       .reply(200, { data: 'Logged in successfully' }); // Mock reponse code and data

//       const expectedActions = [
//         { type: userConstants.LOGIN_SUCCESS }
//       ]

//     const store = mockStore({ })
//     // Dispatch action to login user
//     return store.dispatch(actions.login('12345'))
//       .then(() => { // return of async actions
//         expect(store.getActions()).toEqual(expectedActions);
//       })
//   })

// it('creates LOGIN_FAIL if user login fails', () => {
//   nock('http://localhost:5000')
//     .persist()
//     .get('/config_api/login')
//     .reply(404, { data: { error: 404 } })

//   const expectedActions = [
//     {  type: userConstants.LOGIN_FAIL }
//   ]
//   const store = mockStore({ })

//   return store.dispatch(actions.login('12345'))
//     .then(() => { // return of async actions
//       expect(store.getActions()).toEqual(expectedActions)
//     })
// })

// });

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async actions', () => {
  afterEach(() => {
    fetchMock.restore();
  });

  it('creates FETCH_TODOS_SUCCESS when fetching todos has been done', () => {
    fetchMock.getOnce(`${API_HOST}/config_api/login`, {
      headers: { status: 200 }
    });

    const expectedActions = [
      { type: userConstants.LOGIN_SUCCESS, status: 200 }
    ];
    const store = mockStore({});

    return store.dispatch(actions.login()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
