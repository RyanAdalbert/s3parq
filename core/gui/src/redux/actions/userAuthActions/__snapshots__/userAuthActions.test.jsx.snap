// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates LOGIN_SUCCESS when login authorization gets a 200 response 1`] = `
Array [
  Object {
    "oAuthToken": undefined,
    "type": "LOGIN_ATTEMPT",
  },
  Object {
    "error": [Error: FORBIDDEN],
    "type": "LOGIN_FAIL",
  },
]
`;
